core/shared/src/main/scala/scalaz/zio/stream/Chunk.scala

```diff
@@ -0,0 +1,839 @@
+   */
+  final def foldM[E, S](s: S)(f: (S, A) => IO[E, S]): IO[E, S] =
+    foldLeft[IO[E, S]](IO.now(s)) { (s, a) =>
+      s.flatMap(f(_, a))
+    }
+
+  /**
+   * Folds over the elements in this chunk from the right.
+   */
+  def foldRight[@specialized S](s0: S)(f: (A, S) => S): S = {
+    val len = self.length
+    var s   = s0
+
+    var i = len - 1
+    while (i >= 0) {
+      s = f(self(i), s)
+      i -= 1
+    }
+
+    s
+  }
+
+  override final def hashCode: Int = toArray.toSeq.hashCode
+
+  /**
+   * Returns a chunk with the elements mapped by the specified function.
+   */
+  def map[@specialized B](f: A => B): Chunk[B] = {
+    val len  = self.length
+    var dest = null.asInstanceOf[Array[B]]
+
+    var i = 0
+    while (i < len) {
+      val b = f(self(i))
+
+      if (dest == null) {
+        implicit val B: ClassTag[B] = Chunk.Tags.fromValue(b)
+
+        dest = Array.ofDim[B](len)
+      }
+
+      dest(i) = b
+
+      i = i + 1
+    }
+
+    if (dest != null) Chunk.Arr(dest)
+    else Chunk.Empty
+  }
+
+  /**
+   * Materializes a chunk into a chunk backed by an array. This method can
+   * improve the performance of bulk operations.
+   */
+  def materialize[A1 >: A]: Chunk[A1] = Chunk.Arr(toArray)
+
+  /**
+   * Generates a readable string representation of this chunk using the
+   * specified start, separator, and end strings.
+   */
+  final def mkString(start: String, sep: String, end: String): String = {
+    val builder = new scala.collection.mutable.StringBuilder()
+
+    builder.append(start)
+
+    var i = 0
+    while (i < self.length) {
+      if (i != 0) builder.append(sep)
+      builder.append(self(i).toString)
+      i += 1
+    }
+
+    builder.append(end)
+
+    builder.toString
+  }
+
+  /**
+   * Generates a readable string representation of this chunk using the
+   * specified separator string.
+   */
+  final def mkString(sep: String): String = mkString("", sep, "")
+
+  /**
+   * Generates a readable string representation of this chunk.
+   */
+  final def mkString: String = mkString("")
+
+  /**
+   * Scans over the chunk, statefully producing new elements of type `B`.
+   */
+  final def scan[@specialized S1, @specialized B](s1: S1)(f1: (S1, A) => (S1, B)): (S1, Chunk[B]) = {
+    var s: S1          = s1
+    var i              = 0
+    var dest: Array[B] = null.asInstanceOf[Array[B]]
+
+    while (i < self.length) {
+      val a = self(i)
+      val t = f1(s, a)
+
+      s = t._1
+      val b = t._2
+
+      if (dest == null) {
+        implicit val B: ClassTag[B] = Chunk.Tags.fromValue(b)
+
+        dest = Array.ofDim(self.length)
+      }
+
+      dest(i) = b
+
+      i += 1
+    }
+
+    s ->
+      (if (dest == null) Chunk.empty
+       else Chunk.Arr(dest))
+  }
+
+  /**
+   * Take sthe first `n` elements of the chunk.
+   */
+  final def take(n: Int): Chunk[A] =
+    self match {
+      case Chunk.Slice(c, o, l) =>
+        Chunk.Slice(c, o, l - n - 1)
```

Comment:
`if l - n - 1 <= 0` => Chunk.Empty

Reply:
Probably my fault, I copied from `splitAt` because I realized we could eliminate the tuple allocation by dropping and taking separately.